### PR TITLE
Adds proper_tags to cache

### PR DESCRIPTION
**Pull request by fernandog**
As discussed:

Out of the scope of this PR: Add proper_tags in sickbeard.db | history table

### CODE REVIEW

File: sickbeard/tvcache.py [hunk 1]
```diff
@@ -170,6 +176,13 @@ def _get_hash(self, item):
         """
         return self.provider._get_hash(item)
 
+    def _get_proper_count(self, title):
+        """
+        Return proper_count of the item.
+        """
+        guess = guessit(title)
+        return guess.get('proper_count', 0)
+
```
Comment (fernandog) [inline sickbeard/tvcache.py]: @pymedusa/developers here is where I 'guess'

File: sickbeard/tvcache.py [hunk 2]
```diff
@@ -269,6 +282,7 @@ def _parseItem(self, item):
         size = self._get_size(item)
         pubdate = self._get_pubdate(item)
         torrent_hash = self._get_hash(item)
+        proper_count = self._get_proper_count(title)
```
Comment (fernandog) [inline sickbeard/tvcache.py]: @pymedusa/developers I call the 'guess' here

File: sickbeard/manual_search.py [hunk 3]
```diff
@@ -213,8 +213,7 @@ def get_provider_cache_results(indexer, show_all_results=None, perform_search=No
             # The default sql, that's executed for each providers cache table
             common_sql = b"SELECT rowid, ? AS 'provider_type', ? AS 'provider_image', \
                           ? AS 'provider', ? AS 'provider_id', ? 'provider_minseed', ? 'provider_minleech', \
-                          name, season, episodes, indexerid, url, time, \
-                          CASE WHEN proper_count > 0 THEN 1 ELSE 0 END AS 'isproper', \
+                          name, season, episodes, indexerid, url, time, proper_count, \
```
Comment (fernandog) [inline sickbeard/manual_search.py]: @p0psicles instead of 0|1 we send proper_count

**Comment by p0psicles** [comment]
Before we write another single line of code regarding the propers. We first are going to agree upon a solution.

**Comment by p0psicles** [comment]
I would store proper_tags everywhere as a list. And just before writing to db, join it to a string. And visa/versa.

File: sickbeard/classes.py [hunk 4]
```diff
@@ -209,7 +212,7 @@ def selectSeries(allSeries):
 
 
 class Proper(object):
-    def __init__(self, name, url, date, show, seeders, leechers, size, pubdate, hash):
+    def __init__(self, name, url, date, show, seeders, leechers, size, pubdate, hash, proper_tags):
```
Comment (p0psicles) [inline sickbeard/classes.py]: can hash be changed to torrent_hash here please. As it's a buildin.

File: sickbeard/name_parser/parser.py [hunk 5]
```diff
@@ -121,7 +121,7 @@ def _parse_string(self, name):  # pylint: disable=too-many-locals, too-many-bran
 
         matches = []
         bestResult = None
-        guess = guessit.guessit(name, dict(show_type=self.show_type))
+        guess = guessit.guessit(name, dict(show_type=self.show_type, advanced=True, implicit=True))
         if self.use_guessit:
```
Comment (fernandog) [inline sickbeard/name_parser/parser.py]: @ratoaq2 : this was suggested by @labrys  ( https://thepb.in/p/GZhWNY7gY6LtV#H3-4 )

can we do that in the parser?

@p0psicles

File: sickbeard/manual_search.py [hunk 6]
```diff
@@ -239,7 +239,7 @@ def get_provider_cache_results(indexer, show_all_results=None, perform_search=No
     if combined_sql_q:
         sql_prepend = b"SELECT * FROM ("
         sql_append = b") ORDER BY CASE quality WHEN '{quality_unknown}' THEN -1 ELSE CAST(quality AS DECIMAL) END DESC, " \
-                     b" proper_count DESC, seeders DESC".format(quality_unknown=Quality.UNKNOWN)
+                     b" proper_tags DESC, seeders DESC".format(quality_unknown=Quality.UNKNOWN)
```
Comment (fernandog) [inline sickbeard/manual_search.py]: @p0psicles how to sort here ?

File: sickbeard/properFinder.py [hunk 7]
```diff
@@ -158,7 +158,7 @@ def _get_proper_results(self):  # pylint: disable=too-many-locals, too-many-bran
             # if they haven't been added by a different provider than add the proper to the list
             for proper in cur_propers:
                 guess = guessit(proper.name)
-                if not guess.get('proper_count'):
+                if not guess.get('proper_tags'):
```
Comment (ratoaq2) [inline sickbeard/properFinder.py]: There's no proper_tags in guessit

File: sickbeard/name_parser/rules/rules.py [hunk 8]
```diff
@@ -1829,7 +1829,7 @@ class EnhanceReleaseGroupDetection(Rule):
             "format": "HDTV",
             "video_codec": "h264",
             "other": "Proper",
-            "proper_count": 1,
+            "proper_tags": 1,
```
Comment (ratoaq2) [inline sickbeard/name_parser/rules/rules.py]: There's no proper_tags in guessit

File: sickbeard/name_parser/parser.py [hunk 9]
```diff
@@ -524,7 +524,8 @@ def to_parse_result(self, name, guess):
             'extra_info': ' '.join(ensure_list(guess.get('other'))) if guess.get('other') else None,
             'episode_numbers': ensure_list(guess.get('episode'))
             if guess.get('episode') != guess.get('absolute_episode') else [],
-            'ab_episode_numbers': ensure_list(guess.get('absolute_episode'))
+            'ab_episode_numbers': ensure_list(guess.get('absolute_episode')),
+            'proper_tags': guess.get('proper_tags', 0)
         }
```
Comment (ratoaq2) [inline sickbeard/name_parser/parser.py]: And here

**Comment by fernandog** [comment]
@ratoaq2 @p0psicles did all changes you suggested

File: sickbeard/classes.py [hunk 10]
```diff
@@ -212,7 +212,7 @@ def selectSeries(allSeries):
 
 
 class Proper(object):
-    def __init__(self, name, url, date, show, seeders, leechers, size, pubdate, hash, proper_tags):
+    def __init__(self, name, url, date, show, seeders, leechers, size, pubdate, torrent_hash, proper_tags):
```
Comment (medariox) [inline sickbeard/classes.py]: FYI "torrent_hash" is actually specifically called "info_hash" in the [Bittorrent Protocol Specification](https://wiki.theory.org/BitTorrentSpecification). Maybe we want to use that name as well, instead of a more generic and confusing "torrent_hash".

File: sickbeard/tvcache.py [hunk 11]
```diff
@@ -406,7 +417,7 @@ def listPropers(self, date=None):
         before moving on with hitting the providers.
         """
         cache_db_con = self._get_db()
-        sql = b"SELECT * FROM [{provider_id}] WHERE (name LIKE '%.PROPER.%' OR name LIKE '%.REPACK.%' OR name LIKE '%.REAL.%')".format(provider_id=self.provider_id)
+        sql = b"SELECT * FROM [{provider_id}] WHERE is_proper > 0".format(provider_id=self.provider_id)
```
Comment (p0psicles) [inline sickbeard/tvcache.py]: Uuhhh?

Comment (fernandog) [inline sickbeard/tvcache.py]: forgot about that one. sorry

File: sickbeard/classes.py (hunk 10 above)
Comment (p0psicles) [inline sickbeard/classes.py]: I think there both confusing without the right context.  I've been changing hash to torrent_hash pretty much everywhere else.

File: sickbeard/tvcache.py (hunk 11 above)
Comment (p0psicles) [inline sickbeard/tvcache.py]: Haha np

File: sickbeard/name_parser/parser.py (hunk 5 above)
Comment (ratoaq2) [inline sickbeard/name_parser/parser.py]: Nope. Implicit is already in use and advance will just change what's returned by guessit and it will break the rest of the parser since the other parts are expecting the actual values and not a Match object.

File: sickbeard/properFinder.py [hunk 12]
```diff
@@ -158,7 +158,7 @@ def _get_proper_results(self):  # pylint: disable=too-many-locals, too-many-bran
             # if they haven't been added by a different provider than add the proper to the list
             for proper in cur_propers:
                 guess = guessit(proper.name)
-                if not guess.get('proper_tags'):
+                if not guess.get('proper_count'):
```
Comment (medariox) [inline sickbeard/properFinder.py]: This can be completely removed/replaced now, since we know if it is a proper from the NameParser.

File: sickbeard/classes.py (hunk 10 above)
Comment (medariox) [inline sickbeard/classes.py]: If we replaced it with torrent_hash somewhere else, never mind. It would just be the correct naming.

File: sickbeard/name_parser/parser.py [hunk 13]
```diff
@@ -139,7 +139,7 @@ def _parse_string(self, name):  # pylint: disable=too-many-locals, too-many-bran
                 result = ParseResult(name)
                 result.which_regex = [cur_regex_name]
                 result.score = 0 - cur_regex_num
-                result.proper_tags = [x.raw for x in guess.get('other') if x._value=='Proper']
+                result.proper_tags = [x.raw for x in guess.get('other', '') if x._value=='Proper']
```
Comment (medariox) [inline sickbeard/name_parser/parser.py]: What's the reason behind this change?

File: sickbeard/name_parser/parser.py [hunk 14]
```diff
@@ -525,7 +525,8 @@ def to_parse_result(self, name, guess):
             'extra_info': ' '.join(ensure_list(guess.get('other'))) if guess.get('other') else None,
             'episode_numbers': ensure_list(guess.get('episode'))
             if guess.get('episode') != guess.get('absolute_episode') else [],
-            'ab_episode_numbers': ensure_list(guess.get('absolute_episode'))
+            'ab_episode_numbers': ensure_list(guess.get('absolute_episode')),
+            'proper_count': guess.get('proper_count', 0)
```
Comment (medariox) [inline sickbeard/name_parser/parser.py]: ???

File: sickbeard/tvcache.py [hunk 15]
```diff
@@ -386,13 +391,19 @@ def _addCacheEntry(self, name, url, seeders, leechers, size, pubdate, torrent_ha
             # get version
             version = parse_result.version
 
+            # get proper_tags
+            proper_tags = parse_result.proper_tags
+            
+            # Store proper_tags as proper1|proper2|proper3
+            proper_tags = "|".join(proper_tags)
```
Comment (medariox) [inline sickbeard/tvcache.py]: Why not?
```proper_tags = '|'.join(parse_result.proper_tags)```

File: sickbeard/tvcache.py [hunk 16]
```diff
@@ -35,6 +35,7 @@
 from sickrage.show.Show import Show
 
 
+
```
Comment (medariox) [inline sickbeard/tvcache.py]: Should only be 2 new lines.

File: sickbeard/manual_search.py [hunk 17]
```diff
@@ -209,12 +209,11 @@ def get_provider_cache_results(indexer, show_all_results=None, perform_search=No
 
         # TODO: the implicit sqlite rowid is used, should be replaced with an explicit PK column
         # If table doesn't exist, start a search to create table and new columns seeders, leechers and size
-        if table_exists and 'seeders' in columns and 'leechers' in columns and 'size' in columns:
+        if table_exists and 'seeders' in columns and 'leechers' in columns and 'size' in columns and 'proper_tags' in columns:
```
Comment (medariox) [inline sickbeard/manual_search.py]: This can be simplified with all().

File: sickbeard/name_parser/parser.py (hunk 13 above)
Comment (fernandog) [inline sickbeard/name_parser/parser.py]: result.proper_tags = [x.raw for x in guess.get('other') if x._value=='Proper']
 TypeError: 'NoneType' object is not iterable

File: sickbeard/name_parser/parser.py (hunk 14 above)
Comment (fernandog) [inline sickbeard/name_parser/parser.py]: ?

File: sickbeard/properFinder.py (hunk 12 above)
Comment (fernandog) [inline sickbeard/properFinder.py]: no because we only do the parser before adding to cache. Here we only get the list of propers from provider. Its not like the seeders/leechers/size and etc that we do in all providers code file.

File: sickbeard/name_parser/parser.py (hunk 14 above)
Comment (medariox) [inline sickbeard/name_parser/parser.py]: Why are we still getting proper_count?

Comment (fernandog) [inline sickbeard/name_parser/parser.py]: why not ?

File: sickbeard/properFinder.py (hunk 12 above)
Comment (medariox) [inline sickbeard/properFinder.py]: So why exactly are we checking if it is proper if we already know that it is a proper?
Running guessit twice is not an option.

File: sickbeard/name_parser/parser.py (hunk 14 above)
Comment (medariox) [inline sickbeard/name_parser/parser.py]: If it's not used/needed it shouldn't be added just for fun.

File: sickbeard/tvcache.py [hunk 18]
```diff
@@ -406,7 +416,7 @@ def listPropers(self, date=None):
         before moving on with hitting the providers.
         """
         cache_db_con = self._get_db()
-        sql = b"SELECT * FROM [{provider_id}] WHERE (name LIKE '%.PROPER.%' OR name LIKE '%.REPACK.%' OR name LIKE '%.REAL.%')".format(provider_id=self.provider_id)
+        sql = b"SELECT * FROM [{provider_id}] WHERE proper_tag != ''".format(provider_id=self.provider_id)
```
Comment (p0psicles) [inline sickbeard/tvcache.py]: your using proper_tag and proper_tags?

**Comment by p0psicles** [comment]
Personally don't feel much for keeping an attibute proper_count AND proper_tags. What is the impact if we remove all references to proper_count?

**Comment by fernandog** [comment]
@p0psicles all parsed info is cached. so when we call parse again for release all info is stored so if we need proper_count, info is there. It wont parse it again.

File: sickbeard/name_parser/parser.py [hunk 19]
```diff
@@ -634,6 +637,9 @@ def __str__(self):
         if self.release_group:
             to_return += ' [GROUP: {0}]'.format(self.release_group)
 
+        if self.proper_tags:
+            to_return += ' [PROPER COUNT: {0}]'.format(self.proper_tags)
```
Comment (medariox) [inline sickbeard/name_parser/parser.py]: Either change to "PROPER TAGS" or use len().

File: sickbeard/name_parser/parser.py [hunk 20]
```diff
@@ -638,7 +638,7 @@ def __str__(self):
             to_return += ' [GROUP: {0}]'.format(self.release_group)
 
         if self.proper_tags:
-            to_return += ' [PROPER COUNT: {0}]'.format(self.proper_tags)
+            to_return += ' [PROPER TAGs: {0}]'.format("|".join(self.proper_tags))
```
Comment (fernandog) [inline sickbeard/name_parser/parser.py]: @medariox

**Comment by fernandog** [comment]
@ratoaq2 can you rebase this please? Its conflicting in tests/datasets/tvshows.yml

**Comment by fernandog** [comment]
@pymedusa/developers i tested and its working (adding proper tags to cache)

File: gui/slick/views/snatchSelection.mako [hunk 21]
```diff
@@ -359,7 +359,7 @@
                             <img src="${srRoot}/images/providers/missing.png" width="16" height="16" style="vertical-align:middle;" alt="missing provider" title="missing provider"/> ${hItem["provider"]}
                         % endif
                     </td>
-                    <td align="center">${renderQualityPill(int(hItem["quality"]))}</td>
+                    <td align="center">${renderQualityPill(int(hItem["quality"]))} ${('', '<span class="quality Proper">P</span>')[len(hItem["proper_tags"]) > 0]}</td>
```
Comment (medariox) [inline gui/slick/views/snatchSelection.mako]: Why the len()? Just check for existence:
`('proper', '')[not hItem['proper_tags']]

Comment (labrys) [inline gui/slick/views/snatchSelection.mako]: Better yet:
```python
'proper' if hItem['proper_tags'] else ''
```

**Comment by labrys** [comment]
`proper_count` could be replaced with `len(proper_tags)` with no impact

**Comment by fernandog** [comment]
@pymedusa/developers i think this is ready

**Comment by fernandog** [comment]
@duramato 

image

**Comment by fernandog** [comment]
image
